### DISCUSSION TITLE
Is the browser's window object available at the time of pre-rendering?

### DISCUSSION THREAD

**Discussion by rahu-I** [vercel/next.js#29347] [Help] [answered]
Hello, 
This is a question regarding `pre-rendering`. I'm working on a Project where I need to pre-render a page that has mostly static data. 
The data that I'm getting is in the form of HTML and I need to check if the height of the content rendered by the HTML is greater than a certain size, say 70px, then I need to implement a Read More/Read less functionality to hide and show rest of the content. 

What I wanted to ask is `whether we have access to the window or its children objects` at the time of pre-rendering? If for some reason I need to check the window object to see what device is accessing the page, can I do that? Below is a short example of what I'm trying to achieve. Any help is appreciated. 

```
const About = () => {
  const [isLongIntro, setIsLongIntro] = useState(false)
  const aboutRef = useRef()

  useEffect(() => {
    if (aboutRef.current) {
      const height = aboutRef.current.clientHeight
      if (height > 70) {
        setIsLongIntro(true)
      }
    }
  }, [aboutRef.current])

 console.log(typeof window) //What will happen here?

  return (
    <div>
        <div ref={aboutRef} dangerouslySetInnerHTML={{ __html: '<p>Some HTML</p>'}}>
        </div>
    </div>
  )
}

```

**Comment by awareness481** [comment, marked as answer]
The window object doesn't exist on the server so you will get `undefined` on the server.
You could get the [User-Agent](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) in `getServerSideProps` but you can't really get the size of the browser tab until you are on the client.

**Reply by rahu-I** [reply]
Thanks @awareness481 for the quick response.  I should've specified which pre-rendering method I'm using. I am aware that there are 2 ways: 
1. Static Generation (with or without data) 
2. Server side rendering (which uses getServerSideProps)

I'm using **Static generation with data using `getStaticProps`**. So will the window be undefined in this case as well? Does static generation with data also happen at server side? 

Also, In pre-rendering (any type), when HTML is generated for the page, What happens to the refs attached to any element (like what I've done in my code)? Are they also undefined at the time of static generation and initialised value as the DOM node when they're loaded in a browser?

**Reply by awareness481** [reply]
> Thanks @awareness481 for the quick response. I should've specified which pre-rendering method I'm using. I am aware that there are 2 ways:
> 
> 1. Static Generation (with or without data)
> 2. Server side rendering (which uses getServerSideProps)
> 
> I'm using **Static generation with data using `getStaticProps`**. So will the window be undefined in this case as well? Does static generation with data also happen at server side?
> 
> Also, In pre-rendering (any type), when HTML is generated for the page, What happens to the refs attached to any element (like what I've done in my code)? Are they also undefined at the time of static generation and initialised value as the DOM node when they're loaded in a browser?

During Static Generation (SSG) your page will be generated on the server during the build time (when you run `npm run build`). 

About the refs, since you're using `useRef()` the ref will be `{ current: undefined }` on the server during static generation, but that is fine since React will handle the refs during hydration on the client.

> Does static generation with data also happen at server side?

yep

**Reply by rahu-I** [reply]
Thanks again for the response. Please let me know if I understood the flow correctly:

1. The HTML of page is statically generated on the server. The content of the` About div `(with dangerously set HTML) is completely generated. Ref attached to the `About div` is undefined.
2. When the client is sent the page on request,  after initial render the ref will be attached to the About div. Then the `useEffect` will be called and we will have access to the `About div` can check the ref value to see if the height of the About div is larger than a certain height.

Also is there a way to attach  a reference to a DOM node, like about div here, at server? As you mentioned `useRef` will only be available at client side.
